pkg/reconciler/events/pubsub/pubsub.go

```diff
@@ -59,65 +68,14 @@ type Reconciler struct {
 	receiveAdapterName string
 }
 
-// Check that we implement the controller.Reconciler interface.
-var _ controller.Reconciler = (*Reconciler)(nil)
-
-// Reconcile compares the actual state with the desired, and attempts to
-// converge the two. It then updates the Status block of the CloudPubSubSource resource
-// with the current status of the resource.
-func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
-	// Convert the namespace/name string into a distinct namespace and name
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		logging.FromContext(ctx).Desugar().Error("Invalid resource key")
-		return nil
-	}
-
-	// Get the CloudPubSubSource resource with this namespace/name
-	original, err := r.pubsubLister.CloudPubSubSources(namespace).Get(name)
-	if apierrs.IsNotFound(err) {
-		// The CloudPubSubSource resource may no longer exist, in which case we stop processing.
-		logging.FromContext(ctx).Desugar().Error("CloudPubSubSource in work queue no longer exists")
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	// Don't modify the informers copy
-	pubsub := original.DeepCopy()
+// Check that our Reconciler implements Interface.
+var _ cloudpubsubsourcereconciler.Interface = (*Reconciler)(nil)
 
-	reconcileErr := r.reconcile(ctx, pubsub)
-
-	// If no error is returned, mark the observed generation.
-	if reconcileErr == nil {
-		pubsub.Status.ObservedGeneration = pubsub.Generation
-	}
-
-	if equality.Semantic.DeepEqual(original.Status, pubsub.Status) {
-		// If we didn't change anything then don't call updateStatus.
-		// This is important because the copy we loaded from the informer's
-		// cache may be stale and we don't want to overwrite a prior update
-		// to status with this stale state.
-
-	} else if uErr := r.updateStatus(ctx, original, pubsub); uErr != nil {
-		logging.FromContext(ctx).Desugar().Warn("Failed to update CloudPubSubSource status", zap.Error(uErr))
-		r.Recorder.Eventf(pubsub, corev1.EventTypeWarning, "UpdateFailed",
-			"Failed to update status for CloudPubSubSource %q: %v", pubsub.Name, uErr)
-		return uErr
-	} else if reconcileErr == nil {
-		// There was a difference and updateStatus did not return an error.
-		r.Recorder.Eventf(pubsub, corev1.EventTypeNormal, "Updated", "Updated CloudPubSubSource %q", pubsub.Name)
-	}
-	if reconcileErr != nil {
-		r.Recorder.Event(pubsub, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
-	}
-	return reconcileErr
-}
-
-func (r *Reconciler) reconcile(ctx context.Context, pubsub *v1alpha1.CloudPubSubSource) error {
+func (r *Reconciler) ReconcileKind(ctx context.Context, pubsub *v1alpha1.CloudPubSubSource) pkgreconciler.Event {
 	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("pubsub", pubsub)))
 
 	pubsub.Status.InitializeConditions()
+	pubsub.Status.ObservedGeneration = pubsub.Generation
```

Comment:
The previous behavior was to set ObservedGeneration only after a successful reconcile. This seems like the correct behavior as it indicates that reconciliation is still needed at the current generation.

Reply:
done. ptal